src/Build/Graph/ProjectGraph.cs

```diff
@@ -243,6 +297,92 @@ private static ImmutableList<string> ExpandDefaultTargets(ProjectInstance projec
+        /// <remarks>
+        /// The behavior of this method should match the logic in Microsoft.Common.CurrentVersion.targets and the MSBuild task.
+        /// </remarks>
+        private static PropertyDictionary<ProjectPropertyInstance> GetProjectReferenceGlobalProperties(ProjectItemInstance projectReference, PropertyDictionary<ProjectPropertyInstance> requesterGlobalProperties)
+        {
+            string propertiesString = projectReference.GetMetadataValue(PropertiesMetadataName);
+            string additionalPropertiesString = projectReference.GetMetadataValue(AdditionalPropertiesMetadataName);
+            string undefinePropertiesString = projectReference.GetMetadataValue(UndefinePropertiesMetadataName);
+            string globalPropertiesToRemoveString = projectReference.GetMetadataValue(GlobalPropertiesToRemoveMetadataName);
+
+            // TODO: Mimic AssignProjectConfiguration's behavior for determining the values for these.
+            string setConfigurationString = projectReference.GetMetadataValue(SetConfigurationMetadataName);
+            string setPlatformString = projectReference.GetMetadataValue(SetPlatformMetadataName);
+            string setTargetFrameworkString = projectReference.GetMetadataValue(SetTargetFrameworkMetadataName);
+
+            // The properties on the project reference supercede the ones from the MSBuild task instad of appending.
+            if (string.IsNullOrEmpty(propertiesString)
+                && (!string.IsNullOrEmpty(setConfigurationString) || !string.IsNullOrEmpty(setPlatformString) || !string.IsNullOrEmpty(setTargetFrameworkString)))
+            {
+                propertiesString = $"{setConfigurationString};{setPlatformString};{setTargetFrameworkString}";
+            }
+
+            // If none of these are set, we can just reuse the requestor's global properties directly.
+            if (string.IsNullOrEmpty(propertiesString)
+                && string.IsNullOrEmpty(additionalPropertiesString)
+                && string.IsNullOrEmpty(undefinePropertiesString)
+                && string.IsNullOrEmpty(globalPropertiesToRemoveString))
+            {
+                return requesterGlobalProperties;
+            }
+
+            // Make a copy to avoid mutating the requester
+            var globalProperties = new PropertyDictionary<ProjectPropertyInstance>(requesterGlobalProperties);
+
+            // Append and remove properties as specified by the various metadata
+            MergeIntoPropertyDictionary(globalProperties, propertiesString, PropertiesMetadataName);
+            MergeIntoPropertyDictionary(globalProperties, additionalPropertiesString, AdditionalPropertiesMetadataName);
+            RemoveFromPropertyDictionary(globalProperties, globalPropertiesToRemoveString);
+            RemoveFromPropertyDictionary(globalProperties, undefinePropertiesString);
+
+            return globalProperties;
+        }
+
+        private static void MergeIntoPropertyDictionary(
```

Comment:
Should this be a method (or extension method) on `PropertyDictionary`? I feel like the mutations would read better that way.

Reply:
Unfortunately `PropertyDictionary` is actually `PropertyDictionary<T>` where `T` in this case is `ProjectPropertyInstance`. So a method like that would need to know how to convert stuff to `T` generically.

There is actually already a `ImportProperties<T>`, so I could actually just convert the string to an `IEnumerable<ProjectPropertyInstance>`, but really at that point we're not saving much.

I guess I could create an extension method for `PropertyDictionary<ProjectPropertyInstance>` specifically. That would allow shared code between the MSBuild task and here, but there's still the question about the weird-ish way I'm using `PropertyParser.GetTableWithEscaping` here.

Reply:
Actually, not really any shareable code between this and the MSBuild task, as it uses Dictionary<string, string> while we use `PropertyDictionary<ProjectPropertyInstance>`. So besides a nicer syntax, there's not really much benefit here to having an extension method vs just a static method in this class.